xarray/backends/core.py

```diff
@@ -0,0 +1,425 @@
+        -------
+        Writable array-like object, see `create_variable` for details.
+        """
+        # Note: this mostly exists for the benefit of future support for partial
+        # reads -- we don't actually make use of this in the current version of
+        # xarray.
+        raise NotImplementedError
+
+    def set_attribute(self, name, value):
+        # type: (Hashable, Any) -> None
+        """Set a global attribute on the DataStore."""
+        raise NotImplementedError
+
+    def sync(self):
+        """Synchronize writes to this DataStore."""
+        pass
+
+    def get_write_lock(self, name, region=Ellipsis):
+        # type: (Hashable, Union[Ellipsis, Tuple[slice, ...]]) -> object
+        """Return a lock for writing a given variable.
+
+        This method may be useful for DataStores that from which data is
+        written in parallel (e.g., with dask).
+
+        Parameters
+        ----------
+        name : Hashable
+            Variable name.
+        region : Union[Ellipsis, Tuple[slice, ...]], optional
+            Region in the variable, e.g., valid key argument to `data[key]`,
+            for which to get a lock.
+
+        Returns
+        -------
+        threading.Lock() ducktype (that is, an object with `acquire` and
+        `release` methods), or None, if no lock is necessary.
+        """
+        # Again, we actually have a use for the region argument? Could be useful
+        # to ensure writes to zarr are safe.
+        return None
+
+
+class InMemoryDataStore(AbstractWritableDataStore):
+    """Stores variables and attributes directly in OrderedDicts.
+
+    This store exists internal testing purposes, e.g., for integration tests
+    with dask.array that will not need to write actual data to disk.
+    """
+    def __init__(self):
+        self._variables = OrderedDict()
+        self._attributes = OrderedDict()
+        # do we need locks? are writes to NumPy arrays thread-safe?
+        # this is a dumb but safe approach.
+        self._write_locks = defaultdict(threading.Lock)
+
+    def get_variables(self):
+        return self._variables
+
+    def get_attributes(self):
+        return self._attributes
+
+    def get_read_lock(self, name, region=Ellipsis):
+        return None
+
+    def create_variable(self, name, variable, check_encoding=False):
+        if check_encoding and variable.encoding:
+            raise ValueError('encoding must be empty')
+        store_variable = Variable(variable.dims,
+                                  np.empty_like(variable),
+                                  copy.deepcopy(variable.attrs))
+        self._variables[name] = store_variable
+        return store_variable.values
+
+    def get_writable_array(self, name):
+        return self._variables[name].values
+
+    def set_attribute(self, name, value):
+        self._attributes[name] = copy.deepcopy(value)
+
+    def get_write_lock(self, name, region=Ellipsis):
+        return self._write_locks[name]
```

Comment:
This, presumably, is what @alimanfoo would implement if we wanted to support climate data on Zarr?